gui-agent/vmside.c

```diff
@@ -1626,62 +1658,71 @@ static void handle_keypress(Ghandles * g, XID UNUSED(winid))
+            state.mods &= LockMask;
+            key.state &= LockMask;
+        }
+        if (state.mods != key.state) {
+            XModifierKeymap *modmap;
+            int mod_index;
+            int mod_mask;
+
+            modmap = XGetModifierMapping(g->display);
+            if (!modmap) {
+                if (g->log_level > 0)
+                    fprintf(stderr, "failed to get modifier mapping\n");
+            } else {
+                // from X.h:
+                // #define ShiftMapIndex           0
+                // #define LockMapIndex            1
+                // #define ControlMapIndex         2
+                // #define Mod1MapIndex            3
+                // #define Mod2MapIndex            4
+                // #define Mod3MapIndex            5
+                // #define Mod4MapIndex            6
+                // #define Mod5MapIndex            7
+                for (mod_index = 0; mod_index < 8; mod_index++) {
+                    if (modmap->modifiermap[mod_index*modmap->max_keypermod] == 0x00) {
+                        if (g->log_level > 1)
+                            fprintf(stderr, "ignoring disabled modifier %d\n", mod_index);
+                        // no key set for this modifier, ignore
+                        continue;
+                    }
+                    mod_mask = (1<<mod_index);
+                    // special case for caps lock switch by press+release
+                    if (mod_index == LockMapIndex) {
+                        if ((state.mods & mod_mask) ^ (key.state & mod_mask)) {
+                            feed_xdriver(g, 'K', modmap->modifiermap[mod_index*modmap->max_keypermod], 1);
+                            feed_xdriver(g, 'K', modmap->modifiermap[mod_index*modmap->max_keypermod], 0);
+                        }
+                    } else {
+                        if ((state.mods & mod_mask) && !(key.state & mod_mask))
+                            feed_xdriver(g, 'K', modmap->modifiermap[mod_index*modmap->max_keypermod], 0);
+                        else if (!(state.mods & mod_mask) && (key.state & mod_mask))
+                            feed_xdriver(g, 'K', modmap->modifiermap[mod_index*modmap->max_keypermod], 1);
                     }
-                } else {
-                    if ((state.mods & mod_mask) && !(key.state & mod_mask))
-                        feed_xdriver(g, 'K', modmap->modifiermap[mod_index*modmap->max_keypermod], 0);
-                    else if (!(state.mods & mod_mask) && (key.state & mod_mask))
-                        feed_xdriver(g, 'K', modmap->modifiermap[mod_index*modmap->max_keypermod], 1);
                 }
+                XFreeModifiermap(modmap);
             }
-            XFreeModifiermap(modmap);
         }
-    }
 
-    feed_xdriver(g, 'K', key.keycode, key.type == KeyPress ? 1 : 0);
+        feed_xdriver(g, 'K', key.keycode, key.type == KeyPress ? 1 : 0);
+    } else {
+        struct input_event iev;
```

Comment:
Not syncing modifiers state here looks suspicious. What happens for example if you Alt-tab into another's VM window, and switch back but using mouse? Won't Alt remain stuck? There may be also other similar cases, like opening a menu with Alt-F1 etc, switching workspaces with some key pressed etc.

Reply:
I changed it so that the section in handle_keypress that syncs modifiers will run whether input device creation was requested or not in e6d5e592d04d33361eb48eaa5d314bb17c5d313f